### ISSUE TITLE
fzf users wanted

### ISSUE THREAD

**Issue by alejandrogallo** [papis/papis#334]
I'm preparing a `fzf` interface to be put into core
since it's very simple, but I'd like to have some feedback
and tips from longtime fzf users, if there are any,
maybe then we can create a better interface. Fzf really gives
an unparallel speed to the searching.

**Comment by jghauser** [comment]
I don't really have any great ideas, but [fzf-bibtex](https://github.com/msprev/fzf-bibtexl) is done really well and has a similar use case to papis. Maybe there's something useful in there!

**Comment by alejandrogallo** [comment]
I checked it out, it looks good, right now there is something similar to the tui, with `C-e` for editing and so on, not exactly the same but the code is much simpler and quite frankly `fzf` is much faster, but it can't do multiline, that would be a deal breaker for some, but it looks ok
![image](https://user-images.githubusercontent.com/12715358/121759100-69d8a900-cb24-11eb-835a-43fb66923dca.png)

**Comment by alejandrogallo** [comment]
It works very well, I'll include it as a part of core papis from verion `0.12` if everyone is ok with it.

**Comment by jghauser** [comment]
The lack of multiline is a shame, I do use them in the default interface. But I guess the speed gain and the consistency for existing fzf users like me make up for that. I'm also wondering whether this would make it easier to integrate papis in vim with one of the existing fzf based plugins. I'll definitely check this out in the next days!

**Comment by jghauser** [comment]
Btw, will it allow searching for example tags in particular? And will in allow searching tags even if I disable having them visible in the interface (because of the single line space constraint)?

**Comment by alejandrogallo** [comment]
That's right, I mentioned something about this in the documentation, this is why I added since the very beginning to the 
the `match-format` concept, to be able to match in a hidden way against journal title, year, tags etc
without being overwhelmed by long strings in the headers.

To my knowledge this is not at all possible in fzf, but I have been using fzf for personal
scripts since a week, so I don't know what is possible if one customizes its function
to the extreme. As far as I could understand from the documentation and from the manual page,
it is not really possible. If you or some other fzf user knows a workaround for this,
that would be very useful, as for know I see no way of implementing it.

About better vim interoperability, I guess it would be possible now also with the `fzf.vim`
package, but I am not sure how since I don't use vim for these things anymore and I have no
real motivation of looking into it, but I am sure that someone will take a look at it
eventually.

**Comment by jghauser** [comment]
I just noticed an issue: When there are multiple attached items, pressing enter in the fzf picker opens a second picker (which is great). However, the newline character between entries seems to be missing and fzf just shows one item (with a path that is a concatenation of both files' paths).

**Comment by alejandrogallo** [comment]
That's very true, I just fixed it, sorry about that !

**Comment by jghauser** [comment]
Great, works perfectly now!

Some other things I noticed:

- The ctrl+foo things don't work for me. They only launch once I close the fzf picker.
- Jinja2 formatter doesn't work.
- Because fzf only searches displayed things, the fact that the default `header-format` shortens titles and author list means that often I don't find things. This happens especially with authors because often only the first will be displayed fully. (But this is expected given the fzf limitations we already discussed)

**Comment by alejandrogallo** [comment]
To point 1, yes I know, sady it only works after fzf closes,
right now I don't know how to do it better.

The problem is telling the python process which papis document should be open
or dealt with. Right now I do the following, from fzf I echo the indices
of the list, i.e., If I choose documents 5 8 and 15 to open,
then the fzf process will echo

```
choose 5 8 15
```

and then I recognise, ok it's a `choose` command and I parse the numbers.
But this only gets outputed to stdout once fzf exits, I have to
deal with it having two threads, but I have to come to do it.

- About the jinja2 problem, the default formatting is done for the python
  formatter, but if you put the `fzf-header-format` ina jinja2 valid format
  and you have your `formater=jinja2` and so on it should work right?
  I don't know how jinja2 works, but it should be something like

    ```
    fzf-header-format = {{doc[title]}} ....
    ```
  instead of
    ```
    fzf-header-format = {doc[title]} ....
    ```

**Comment by jghauser** [comment]
My bad! Didn't read carefully enough and thought that the existing `header-format` was used. With `fzf-header-format` it does indeed work! What doesn't work, however, is markup. Color tags (e.g. `<ansigreen>foo</ansigreen>`) and font style (e.g. `<b>foo</b>`) show up as such rather than changing the text. I'll have a look as to why.

About the the other thing. Did you check out fzf command execution? A quick look at the `man` file tells me the following. Looks like it might do what we need (in particular, 'can execute arbitrary commands without leaving fzf')?

     COMMAND EXECUTION
         With execute(...) action, you can execute arbitrary commands without leaving fzf.  For  example,
         you can turn fzf into a simple file browser by binding enter key to less command like follows.
  
             fzf --bind "enter:execute(less {})"
  
         You can use the same placeholder expressions as in --preview.
  
         fzf  switches  to  the alternate screen when executing a command. However, if the command is ex‐
         pected to complete quickly, and you are not interested in its output, you might want to use exe‐
         cute-silent  instead,  which  silently executes the command without the switching. Note that fzf
         will not be responsive until the command is complete. For  asynchronous  execution,  start  your
         command as a background process (i.e. appending &).
  
         On  *nix  systems, fzf runs the command with $SHELL -c if SHELL is set, otherwise with sh -c, so
         in this case make sure that the command is POSIX-compliant.

**Comment by alejandrogallo** [comment]
About the colors and stuff, you'll have to deal with ansi codes,
you can look at the
[docs](https://papis.readthedocs.io/en/latest/configuration.html#config-settings-fzf-header-format)
and maybe if it is not clear you can tell me.

Here is a list of [ansi codes](https://en.wikipedia.org/wiki/ANSI_escape_code#Colors)

The `<red>` and these goodies where from the formatting of `prompt-toolkit`
provided functionality, fzf does not provide such a parser. We could improve this,
but on the other hand I think ansi codes should also work, if you think it's too
much for the average user we can wrap it somehow.

**Comment by jghauser** [comment]
Oh thanks, I somehow thought these were part of jinja2 and not prompt-toolkit. It now works!

I think it's quite well explained in the documentation. The only confusing thing is that at one point you say what's needed is `\x1b[31m` and at another `[31m`. I guess `\x1b` is how your editor shows the escape code, but that's not how it looks in mine (vim). The second way of writing it hides the escape code in html completely. However, when I copy and paste the second thing into my text editor, it works (displays as `^[[33m` in vim). I think html simply doesn't show the escape code. Maybe a note about this would help some people.

**Comment by jghauser** [comment]
If I'm not mistaken (and I might very well be!), it's not FZF that stdouts the echo commands only when it completes but popen that only reads them then. So, doing something like [this](https://www.endpoint.com/blog/2015/01/28/getting-realtime-output-using-python) should allow you to get the stdout before the command completes.

**Comment by alejandrogallo** [comment]
Hi!

- About the color, you're right!
    The config reader is not parsing `\x1b` as an escape sequence but
    as the string `\x1b` which will be a problem with people that cannot
    add pure ASCII characters from their editors.

    So I changed the formatter to add the `c` variable to contain the colorama
    package and then the users can add the colors in a lexically
    amenable form.

    I changed the documentation about that so you should also change your format,
    to wit, red will be now
    ```
    {c.Fore.RED}
    ```
    etc..

    Thanks for the comment!
    
- Concerning the commands, I have not been able to fix it yet, I should be able
  to do it at some point. I'll leave it as a todo and maybe also someone can hack
  something into it. I found a workaround but it turned out to be quite bad.
  
- The problem about matching the pattern with the whole title I thought about some
  way of finding a compromise and a workaround, of course every user
  can then change it at will but maybe it would be the best behaviour
  as a default.
  Namely, what about having the whole string with a separator and having the preview to show the entry that we have selected to really read more, something like this:,
![image](https://user-images.githubusercontent.com/12715358/122657359-2e149380-d163-11eb-8057-e82ed28ba16e.png)

  and of course the user might change the whole fzf interaction, but as a default, maybe this would be better.

**Comment by jghauser** [comment]
I just tried the colorama colours, but they don't seem to work with jinja2. The codes just turn up as text in the fzf picker. Am I doing something wrong or is this a limitation? I think it's ok as a limitation as jinja2 is an advanced option anyway. Making a note about this somewhere would be good though.

I'm gonna have a look at the commands thing as well. While I was a bit hesitant about losing multiline in the beginning, I now really like the fzf picker and use it as default. I'd be really interested in getting this to work. I'd also like to implement fzf's `reload` function, so that I can reload the list of items without exiting fzf. My hope is to be able to have the fzf picker constantly open and search and open files as needed from it without ever exiting it. I guess it's a habit I have from my earlier zotero use. It would also help mitigate the fact that jinja2 is quite slow, so opening the picker usually takes a few seconds. I'll get back to you if I can make anything work. What approach did you try and then abandon?

Finally, about searching long titles and author lists. Your suggestion is to simply add the complete string or to have the same as before but add the complete strings far to the right so that they wouldn't (ordinarily) be visible (until searching)? I'm doing something like the first, and with jinja2, it's easily implementable. My problem without jinja2 is getting rid of extraneous fields (e.g. to only use surnames) and to deal with special cases (e.g. editors in case of collections etc.). Without removing these, often the title might not be visible etc. I think, adding a preview window would be a good default setting to mitigate this. I think being able to find the documents with any of the words in the fields that are shown is important and what I would expect as a user.

In case you want to mention a jinja2 config as an alteranative in the docs, this works very well for me:

```
fzf-header-format = [33m{%% if doc['author'] %%}{%% for i in doc['author_list'] %%}{{i.get('family')}}{%% if not loop.last %%},{%% endif %%} {%% endfor %%}{%% elif doc['editor'] %%} {{doc['editor'].split(',')[0]}} {%% endif %%}[0;34m({{doc['year']}}) [37m{{doc['title']}} [0;35m({{doc['type']}}) [36m{{doc['tags']}}
```

And, unrelatedly, already a while back I created a `papis-git` aur package to install the git version of papis in Arch. This complements the existing `papis` package that is based on the released version. Just mentioning this in case you want to add a comment to the docs.

**Comment by alejandrogallo** [comment]
Hi Julian,

- I don't know, right now I tested the colorama colors with jinja like so
    ```
    fzf-header-format = {{c.Fore.MAGENTA}}{{doc["title"]}}{{c.Style.RESET_ALL}} :-: {{c.Fore.CYAN}}{{doc["author"]}}{{c.Style.RESET_ALL}} :-:   {{c.Fore.YELLOW}}«{{doc["year"]}}»{{c.Style.RESET_ALL}} :-:   +{{doc["tags"]}}
    ```
    and it seemed to work.


- About the full string search yes that is exactly what I did. Since everythink is implementable from the user perspective I think
  I will simply document the workflow on the documentation and then the users can decide what to do with this idea as an example.

- About the `reload`, that would be cool, hmmm, I am not sure, maybe this is a use case for the papis server.
  I am working from time to time on the webapp for papis, since I really like purescript and really hate javascript.
  For this I coded a simple papis API in the form of a server. Maybe one could simply do a
  `papis fzf` command in a way as a tui application using the server in the background.
  We can talk about this at some point, since I think that otherwise it would be a little difficult, but maybe still doable.
  I'll also think about this.
  Let me get back to you with the things I tested in this sense.

- About `papis-git`, that is excellent, I'll update the documentation right now mentioning you too! Thanks for that!

**Comment by jghauser** [comment]
Oh, my bad. I put the colorama tags outside brackets just like the escape codes. Clearly I don't really know how jinja2 parsing works. Thanks, works perfectly now!

Great, papis server sounds like a cool idea! I'm definitely interested in discussing this and thinking about how to make it work. I've been wanting to look at closer look at the overall codebase, so I guess this would be a great opportunity to do so :).

The approach I was thinking of for opening and editing without leaving fzf was to have some sort of helper script that gets called by fzf which looks up the file associated with a given entry (maybe identified by `ref`) and open it. Of course that would not be all that elegant (and wouldn't make the reload thing work).

**Comment by jghauser** [comment]
I just saw your new alternative config for fzf -- I really like it! I think there might be a small mistake in it, at least for me it only started working once I doubled the % so that is says `"bottom:wrap:15%%"`.

**Comment by alejandrogallo** [comment]
Yes you're completely right, I did not think about that, `%` is a reserverd symbol for the `configparser` string interpolator, so it should be escaped by writing two `%`, I'll fix it right away.
Yes, I also like it, but I was unsure if it is too much for most users, so it is documented at least ;)
Thanks for the comment!

**Comment by rhaynes74** [comment]
Hi folks, I am wondering how to instigate a papis fzf "session"?  For example, should papis list --all bring up a fzf dialog?

**Comment by alexfikl** [comment]
> Hi folks, I am wondering how to instigate a papis fzf "session"? For example, should papis list --all bring up a fzf dialog?

You need to set `picktool = fzf` in your config (at `~/.config/papis/config` on Linux) and then
```
papis list <query>
papis open <query>
```
and others where you need to pick a result will show a `fzf` based dialog. There's some documentation at https://papis.readthedocs.io/en/latest/configuration.html#fzf-integration. 

Does that not work for you?

**Comment by rhaynes74** [comment]
Thanks - I was missing the piece about including a query which returned multiple hits, I was hoping to be able to list all documents in the library.

A related question - I have documents which have an added notes file, is there anyway to get all files included in a document returned from a query?

Sincerely,

Dr. Ronald D. Haynes
Professor, Department of Mathematics and Statistics
Chair, MSc and Phd Scientific Computing Programs
Memorial University of Newfoundland

We acknowledge that the lands on which Memorial University’s campuses are situated are in the traditional territories of diverse Indigenous groups, and we acknowledge with respect the diverse histories and cultures of the Beothuk, Mi’kmaq, Innu, and Inuit of this province.

**Comment by alexfikl** [comment]
> Thanks - I was missing the piece about including a query which returned multiple hits, I was hoping to be able to list all documents in the library.

@rhaynes74 Does going just `papis list` not pull up `fzf` with all the documents in the library? Otherwise, you can use a bit of regex in there, e.g. for the default `database-backend = papis`
```
papis list '.'
```
should also select all documents.

> A related question - I have documents which have an added notes file, is there anyway to get all files included in a document returned from a query?

Yeah, sure!
```
papis list --file <query>
papis list --notes <query>
```
should list the files (usually PDFs) or notes from the selected documents. For notes, you can also use
```
papis edit --notes <query>
```
to open the notes.

**Comment by rhaynes74** [comment]
Thanks!  See comments below:



We acknowledge that the lands on which Memorial University’s campuses are situated are in the traditional territories of diverse Indigenous groups, and we acknowledge with respect the diverse histories and cultures of the Beothuk, Mi’kmaq, Innu, and Inuit of this province.
On Mar 23, 2023 at 6:54 AM -0230, Alex Fikl ***@***.***>, wrote:
		Thanks - I was missing the piece about including a query which returned multiple hits, I was hoping to be able to list all documents in the library.
@rhaynes74 Does going just papis list not pull up fzf with all the documents in the library? Otherwise, you can use a bit of regex in there, e.g. for the default database-backend = papis
papis list '.'
should also select all documents.

Neither works - I am using the whoosh backend if that matters (at least I have it specified in the config file).

Thanks!
		A related question - I have documents which have an added notes file, is there anyway to get all files included in a document returned from a query?
Yeah, sure!
papis list --file <query>
papis list --notes <query>
should list the files (usually PDFs) or notes from the selected documents. For notes, you can also use
papis edit --notes <query>
to open the notes.

**Comment by alexfikl** [comment]
> I am using a whoosh backend. Papis list ‘.’ Doesn’t work to get a list of all documents. 

Ah, for `whoosh` this should work instead:
```
papis list '*'
```
The `whoosh` and the `papis` backends have different search syntax.

> I was hoping to be able to get a list of all files/notes (at the same time) associated with a document.  Is that possible?

That's not possible at the moment, as far as I know, because the files and notes are stored differently. You can hack around it by making a little function in bash (not tested, but something like this should work)
```
def papis-all-files() {
	local papis_id=$(papis list --id $1)
	papis list --files -a "papis_id:$papis_id"
	papis list --notes -a "papis_id:$papis_id"
}
```
Feel free to open a separate bug report for this though!

**Comment by mbrunnen** [comment]
`papis -s picktool fzf open` works fine on Version 0.13. Shouldn't this issue here be closed?

**Comment by alexfikl** [comment]
> `papis -s picktool fzf open` works fine on Version 0.13. Shouldn't this issue here be closed?

Yeah, any additional problems can probably be made into new issues. @alejandrogallo What do you think?
